include/ceed/types.h

```diff
@@ -13,6 +13,23 @@
 #include <stddef.h>
 #include <stdint.h>
 
+/**
+ This macro defines compiler attributes to the CEED_QFUNCTION to force inlining
+ for called functions. The `inline` declaration does not necessarily enforce a
+ compiler to inline a function. This can be deterimental to performance, so
+ here we force inlining to occur unless inlining has been forced off (like
+ during debugging).
+**/
+#ifndef __NO_INLINE__
+#  if defined(__GNUC__) || defined(__clang__)
+#    define CEED_QFUNCTION_ATTR __attribute__((flatten))
+#  else
+#    define CEED_QFUNCTION_ATTR
+#  endif
+#else
+#  define CEED_QFUNCTION_ATTR
+#endif
```

Comment:
I think we want to let backends override `CEED_QFUNCTION_ATTR`

```suggestion
/**
  @ingroup CeedQFunction
  This macro defines compiler attributes to the CEED_QFUNCTION to force inlining
    for called functions. The `inline` declaration does not necessarily enforce a
    compiler to inline a function. This can be deterimental to performance, so
    here we force inlining to occur unless inlining has been forced off (like
    during debugging).
**/
#ifndef CEED_QFUNCTION_ATTR
#ifndef __NO_INLINE__
#  if defined(__GNUC__) || defined(__clang__)
#    define CEED_QFUNCTION_ATTR __attribute__((flatten))
#  else
#    define CEED_QFUNCTION_ATTR
#  endif
#else
#  define CEED_QFUNCTION_ATTR
#endif
#endif
```